Add tests for `nullable` type

sonasingh46 opened the pull request:
We have spun out a separate package, `oapi-codegen/nullable` as a step
towards deepmap/oapi-codegen#1039.

Until we have implemented #27, we cannot add an explicit type alias in
this package, so we can at least add some tests to cover additional
functionality and expectations that the package should have when
interplaying with `oapi-codegen`.

Co-authored-by: Sebastien Guilloux <sebastien.guilloux@elastic.co>
Co-authored-by: Ashutosh Kumar <ashutosh.kumar@elastic.co>

jamietanna commented on types/nullable.go:
```diff
@@ -0,0 +1,31 @@
+
+// Optional type which can help distinguish between if a value was explicitly
+// provided in JSON or not
+type Optional[T any] struct {
```
Did we not say we'd call this `Nullable`? As `Optional` may mean a non-required field

sonasingh46 replied:
done

jamietanna commented on types/nullable.go:
```diff
@@ -0,0 +1,54 @@
+// MarshalJSON implements the  Marshaler interface.
+func (t Nullable[T]) MarshalJSON() ([]byte, error) {
+	if t.IsNull() {
+		return nullBytes, nil
```
We need to replace this with:
```suggestion
		return []byte("null"), nil
```

Right now, we've got the risk that someone could do something like:

```go
func TestF(t *testing.T) {
	var n Nullable[string]
	n.Null = true
	b, err := n.MarshalJSON()
	fmt.Printf("b: %v\n", b)
	fmt.Printf("b: %s\n", b)
	fmt.Printf("err: %v\n", err)

	b[1] = '_'

	b_, err := n.MarshalJSON()
	fmt.Printf("b: %v\n", b_)
	fmt.Printf("b: %s\n", b_)
	fmt.Printf("err: %v\n", err)

}
```

Which outputs:

```
b: [110 117 108 108]
b: null
err: <nil>
b: [110 95 108 108]
b: n_ll
err: <nil>
```

Note that `nullBytes` from the package is changed.

We could alternatively perform a protective copy, or in this case, just return a literal

sonasingh46 replied:
Did not grok this in the first read. Trying to understand it.

sonasingh46 replied:
Ah! Makes sense now.. So basically some one can alter the nullByte. Good catch.

jamietanna commented on types/nullable_test.go:
```diff
@@ -0,0 +1,248 @@
+			assert.NoError(t, err)
+			assert.Equalf(t, tt.wantNull, obj.Name.IsNull(), "IsNull()")
+			assert.Equalf(t, tt.wantSet, obj.Name.IsSet(), "IsSet()")
+			fmt.Println(obj.Name.Get())
```
```suggestion
```

jamietanna commented on types/nullable.go:
```diff
@@ -0,0 +1,54 @@
+	return t.Set
+}
+
+func (t *Nullable[T]) Get() (value T, null bool) {
```
Think this should be:

```suggestion
func (t *Nullable[T]) Get() (value *T, set bool) {
```

sonasingh46 replied:
On similar line -- now I am thinking we do not need this Get() method too.. Especially when I used in my own implementation -- I did not feel them need to use it especially when it returns 2 values, the code becomes little clumsy.

I think we should add these util functions as and when we need them. wdyt?

jamietanna commented on types/nullable.go:
```diff
@@ -0,0 +1,54 @@
+	return json.Marshal(t.Value)
+}
+
+// IsNull returns true if the value is explicitly provided `null` in json
+func (t *Nullable[T]) IsNull() bool {
+	return t.Null
+}
+
+// IsSet returns true if the value is provided in json
+func (t *Nullable[T]) IsSet() bool {
+	return t.Set
+}
```
Unsure if these are needed if we export the values from the struct?

sonasingh46 replied:
Hmm.. Agreed.
- Not exporting the `Set` and `Null` keys is not a good idea as to create a Nullable type it is just easy to have them.
- Now, since we have the fields exported the methods really does not do any useful stuff. 

I was thinking on that and figured we should have something like the following: 

```go
// HasValue returns true if the value is provided in json
func (t *Nullable[T]) HasValue() bool {
	return t.Set && !t.Null
}
```

HasValue() is a good util to have as I felt the need to that while using this type to answer this question: 
- Is the value provided in JSON as well as is not null. 

If this util is not present then one will have to do: 
```go
if t.Set && !t.Null {// do something}
```
vs
```go
if t.HasValue() {// do something}
```

jamietanna commented on types/nullable_test.go:
```diff
@@ -0,0 +1,248 @@
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t1 *testing.T) {
```
```suggestion
		t.Run(tt.name, func(t *testing.T) {
```

Otherwise we're not passing in the correct type.

Shadowing here is fine (in my opinion - happy to chat if there are things to be aware of)

sonasingh46 replied:
typo.. thanks. 
Will fix that.

jamietanna commented on types/nullable_test.go:
```diff
@@ -0,0 +1,248 @@
+			assert: func(obj TestComplex, t *testing.T) {
+				assert.Equalf(t, true, obj.StringList.IsSet(), "string_list should be set")
+				assert.Equalf(t, false, obj.StringList.IsNull(), "string_list should not be null")
+				gotStringList, isNull := obj.StringList.Get()
```
Mind adding newlines for readability?
```suggestion

				gotStringList, isNull := obj.StringList.Get()
```

jamietanna commented on types/nullable_test.go:
```diff
@@ -0,0 +1,248 @@
+		t.Run(tt.name, func(t1 *testing.T) {
+			var obj SimplePointerInt
+			err := json.Unmarshal(tt.jsonInput, &obj)
+			assert.NoError(t, err)
```
May be worth replacing this and similar ones with:
```suggestion
			require.NoError(t, err)
```

So we're failing earlier

jamietanna commented on types/nullable.go:
```diff
@@ -0,0 +1,54 @@
+
+// MarshalJSON implements the  Marshaler interface.
+func (t Nullable[T]) MarshalJSON() ([]byte, error) {
+	if t.IsNull() {
```
Should this be?
```suggestion
	if t.IsSet() && t.IsNull() {
```

jamietanna replied:
Or maybe returning early if it's not been set?

sonasingh46 replied:
So looks like returning early if not set is bit tricky: 

This causes `unexpected end of JSON input` error
```go
if !t.Set {
		return []byte{}/* or nil*/, nil
	}
```

Or the second option is :
```go
if !t.Set {
		return nullBytes, nil
	}
```
Which looks incorrect. Thinking.

sonasingh46 replied:
Looks worth to go via `if t.IsSet() && t.IsNull() ` that you suggested without overly complicating things.

jamietanna commented on types/nullable_test.go:
```diff
@@ -0,0 +1,248 @@
+	"github.com/stretchr/testify/assert"
+	"testing"
+)
+
```
I think it's also worthwhile having some testable examples, i.e.

```go
func ExampleNullable_unmarshal() {
	obj := struct {
		Name Nullable[string] `json:"name"`
	}{}

	err := json.Unmarshal([]byte(`
		{
		}
		`), &obj)
	if err != nil {
		fmt.Printf("Error: %v\n", err)
		return
	}
	fmt.Printf("obj.Name.Set: %v\n", obj.Name.Set)
	fmt.Printf("obj.Name.Value: %v\n", obj.Name.Value)
	fmt.Println("---")

	err = json.Unmarshal([]byte(`
		{
		"name": null
		}
		`), &obj)
	if err != nil {
		fmt.Printf("Error: %v\n", err)
		return
	}
	fmt.Printf("obj.Name.Set: %v\n", obj.Name.Set)
	fmt.Printf("obj.Name.Value: %v\n", obj.Name.Value)
	fmt.Println("---")

	err = json.Unmarshal([]byte(`
		{
		"name": "foo"
		}
		`), &obj)
	if err != nil {
		fmt.Printf("Error: %v\n", err)
		return
	}
	fmt.Printf("obj.Name.Set: %v\n", obj.Name.Set)
	if obj.Name.Value == nil {
		fmt.Println("Error: expected obj.Name.Value to have a value, but was <nil>")
		return
	}
	fmt.Printf("obj.Name.Value: %#v\n", *obj.Name.Value)
	fmt.Println("---")

	// Output:
	// obj.Name.Set: false
	// obj.Name.Value: <nil>
	// ---
	// obj.Name.Set: true
	// obj.Name.Value: <nil>
	// ---
	// obj.Name.Set: true
	// obj.Name.Value: "foo"
	// ---
}
```

sonasingh46 replied:
Makes sense. 

Since `Value` is not a pointer. The following does not apply.

```go
// when it's set explicitly to nil
	obj.ID.Value = nil
	obj.ID.Set = true
```

jamietanna commented:
Was the original implementation of the code adapted from https://github.com/guregu/null ? It looks like it but we've not declared so, as the licence would expect?

sonasingh46 commented:
>Was the original implementation of the code adapted from https://github.com/guregu/null ? It looks like it but we've not declared so, as the licence would expect?

Not really but definitely adapted the `nullByte` stuff from here. https://github.com/guregu/null/blob/master/string.go#L15
Also I looked at a lot of libs around doing the Nullable distinction and the basic way is kind of same. 

So if we need to declare it or do anything needful, sure! Thanks for bringing this.

jamietanna commented on types/nullable.go:
```diff
@@ -0,0 +1,84 @@
+	"fmt"
+)
+
+type OptionalNullable[T any] struct {
```
We should flip the order of these, so `Nullable` is first

jamietanna commented on types/nullable.go:
```diff
@@ -0,0 +1,84 @@
+// provided `null` in JSON or not
+type Nullable[T any] struct {
+	Value T
+	// Set indicates whether the underlying Value was sent in the request TODO
+	Set bool
```
Shouldn't be needed - it's a required field, so we should assume that `Set` is true, for `Nullable`

Commit:
Add tests for `nullable` type

We have spun out a separate package, `oapi-codegen/nullable` as a step
towards deepmap/oapi-codegen#1039.

Until we have implemented #27, we cannot add an explicit type alias in
this package, so we can at least add some tests to cover additional
functionality and expectations that the package should have when
interplaying with `oapi-codegen`.

Co-authored-by: Sebastien Guilloux <sebastien.guilloux@elastic.co>
Co-authored-by: Ashutosh Kumar <ashutosh.kumar@elastic.co>

sonasingh46 committed:
refactor unit tests

Signed-off-by: Ashutosh Kumar <ashutosh.kumar@elastic.co>

sonasingh46 committed:
add unit test for complex struct

Signed-off-by: Ashutosh Kumar <ashutosh.kumar@elastic.co>

jamietanna commented on types/nullable_test.go:
```diff
@@ -437,8 +466,61 @@ func TestNullableRequired_UnmarshalJSON(t *testing.T) {
 	}
 }
+
+// Idempotency tests for nullable and optional
```
Are these actually `Idempotency` tests? Or can we find a better name?

jamietanna commented on types/nullable_test.go:
```diff
@@ -437,8 +466,61 @@ func TestNullableRequired_UnmarshalJSON(t *testing.T) {
+
+	var obj2 StringNullableOptional
+	originalJSON2 := []byte(`{"id":"12esd412"}`)
+	err2 := json.Unmarshal(originalJSON2, &obj2)
```
Why not re-use `err`?

jamietanna commented on types/nullable_test.go:
```diff
@@ -385,50 +412,52 @@ func TestNullableRequired_UnmarshalJSON(t *testing.T) {
 			name: "when explicitly set to a specific value",
 			json: []byte(`{"replica_count":5}`),
-			assert: func(t *testing.T, obj SimpleIntNullableRequired) {
+			assert: func(obj SimpleIntNullableRequired, t *testing.T) {
```
:stop_sign: please revert this back, the ordering was changed on purpose

jamietanna commented on types/nullable_test.go:
```diff
@@ -385,50 +412,52 @@ func TestNullableRequired_UnmarshalJSON(t *testing.T) {
 				t.Helper()
 
-				assert.Truef(t, obj.ReplicaCount.IsNull(), "replica count should be null")
+				assert.Equalf(t, true, obj.ReplicaCount.IsSpecified(), "replica count should be set")
```
Please revert this, we're using `Truef` on purpose

jamietanna commented on types/nullable_test.go:
```diff
@@ -471,6 +471,143 @@ func TestNullableRequired_UnmarshalJSON(t *testing.T) {
+	type testCase struct {
+		name          string
+		jsonInput     []byte
+		assert        func(obj ComplexNullable, t *testing.T)
```
```suggestion
		assert        func(t *testing.T, obj ComplexNullable)
```
